File: 0271. Encode and Decode Strings.java
// Medium
// String
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/encode-and-decode-strings/

import java.util.*;

class Codec {

    // Encodes a list of strings to a single string.
    // length#string
    public String encode(List<String> strs) {
        StringBuilder sb = new StringBuilder();
        for (String s : strs) {
            sb.append(s.length()).append("#").append(s);
        }
        return sb.toString();
    }

    // Decodes a single string to a list of strings.
    // Get the len and the substring 
    public List<String> decode(String s) {
        List<String> res = new ArrayList<>();
        int i = 0;
        while (i < s.length()) {
            int j = i;
            while (s.charAt(j) != '#') j++;
            int len = Integer.parseInt(s.substring(i, j));
            res.add(s.substring(j + 1, j + 1 + len));
            i = j + 1 + len;
        }
        return res;
    }
}

File: 0001-0500/0271. Encode and Decode Strings.py
# Time:O(n), Space:O(n)

from typing import List

class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string."""
        res = []
        for s in strs:
            res.append(str(len(s)) + ":" + s)
        return "".join(res)

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings."""
        res = []
        i = 0
        while i < len(s):
            j = i
            while s[j] != ":":
                j += 1
            length = int(s[i:j])
            i = j + 1  # skip ':'
            res.append(s[i : i + length])
            i += length

        return res
